File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parantheses_map = {
            "]": "[",
            "}": "{",
            ")": "("
        }
        stack = []

        for char in s:
            if char not in parantheses_map:
                stack.append(char)
            else:
                if not stack:
                    return False
                
                val = stack.pop()

                if val != parantheses_map[char]:
                    return False
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if( s.length === 1 )
        return false
    
    if( s.length === 0 )
        return true

    let openingBrackts =  new Set(["{", "(", "[" ]);
    let c2oMapping = {
        ")": "(",
        "]": "[",
        "}": "{",
    }
    let stack = [];

    for( let char of s ) {
        if( openingBrackts.has(char) )
            stack.push(char);
        else {
            if(  stack[stack.length-1] !== c2oMapping[char] )
                return false
            else
                stack.pop();
        }
    }

    return stack.length === 0;
};

/*

if opening
stack push

if closing 
stack top = oppp if current
stack pop

stack empty

*/